### ISSUE TITLE
base tag ignored for HTMLImports

### ISSUE THREAD

**Issue by garysoed** [webcomponents/webcomponentsjs#329]
E.g.:
```html
<base href="base">

<link rel="import" href="a/b/c.html">
```
The example above tries to load `a/b/c.html` instead of `base/a/b/c.html`

**Comment by jmesserly** [comment]
@sorvell may have thoughts on this one?

fwiw, I don't see anything in [HTML Imports spec](http://www.w3.org/TR/html-imports/) specifically addressing the [base element](https://html.spec.whatwg.org/multipage/semantics.html#the-base-element), but I think it falls out of existing HTML spec language for [document-base-url](https://html.spec.whatwg.org/multipage/infrastructure.html#document-base-url).

I think the issue is that HTML Imports polyfill always creates a `<base>` element in attempt to get the normal URL resolution behavior, and adds it first in `<head>`
https://github.com/webcomponents/webcomponentsjs/blob/c15e80fedc2322c6ac0eccaa7e8501a23ee9f7af/src/HTMLImports/importer.js#L132

Maybe it should have some logic to only add that if there isn't a 'base' already?

I might be missing something obvious here though, not an expert on the imports code :)

**Comment by drarmstr** [comment]
I ran into this problem also.  As motivation: I was using the <base> tag in a web component so I could use <a> links in the component with just a hash anchor path for a router.  Because the links are in a component that is not in the root directory, the browser would append the path to the component when following the link which is obviously incompatible with a router.  While a solution might have been to use "/#path" for the links this didn't work in my case because with "/#..." browsers also clear the current URL query string.  Using <base href="../../"/> in a component seems like a great solution to allow use of hash paths in links in components.

This does in fact work great in Chrome.  However, it seems that the <base> tag is not respected in this Polyfill creating a compatibility problem...

Example:
index.html:

    <!DOCTYPE html>
    <html lang="en">
    <head>
        <meta charset="utf-8" />
        <script src="components/webcomponentsjs/webcomponents-lite.js"></script>    
        <link rel="import" href="components/foo/foo.html" />
    </head>

    <body>
        Main body
    </body>
    </html>
    
components/foo/foo.html:

    <base href="../../" />
    <link rel="import" href="components/bar/bar.html" />
    <script>alert("foo");</script>
    <template>
        <a href="#path/for/router">link</a>
    </template>

components/bar/bar.html:

    <script>alert("bar");</script>

In Chrome the "bar" alert will show up properly, but it will not with the Polyfill.

**Comment by drarmstr** [comment]
Just checking if this bug is on the radar or if we should be looking for a workaround?

**Comment by jmesserly** [comment]
@drarmstr I'd say it's more of an unimplemented feature. Always happy to take a look at patches. The first thing I'd try is change the line I linked above:
https://github.com/webcomponents/webcomponentsjs/blob/c15e80fedc2322c6ac0eccaa7e8501a23ee9f7af/src/HTMLImports/importer.js#L132

to only add `<base>` if it's not already present, and see if that fixes the issue for you

**Comment by drarmstr** [comment]
@jmesserly 
Thank you for the pointer!  Here is the original `makeDocument()`:

    function makeDocument(resource, url) {
      // create a new HTML document
      var doc = document.implementation.createHTMLDocument(IMPORT_LINK_TYPE);
      // cache the new document's source url
      doc._URL = url;
      // establish a relative path via <base>
      var base = doc.createElement('base');
      base.setAttribute('href', url);
      // add baseURI support to browsers (IE) that lack it.
      if (!doc.baseURI && !hasBaseURIAccessor(doc)) {
        // Use defineProperty since Safari throws an exception when using assignment.
        Object.defineProperty(doc, 'baseURI', {value:url});
      }
      // ensure UTF-8 charset
      var meta = doc.createElement('meta');
      meta.setAttribute('charset', 'utf-8');

      doc.head.appendChild(meta);
      doc.head.appendChild(base);
      // install html
      doc.body.innerHTML = resource;
      // TODO(sorvell): ideally this code is not aware of Template polyfill,
      // but for now the polyfill needs help to bootstrap these templates
      if (window.HTMLTemplateElement && HTMLTemplateElement.bootstrap) {
        HTMLTemplateElement.bootstrap(doc);
      }
      return doc;
    }

And here is a modified `makeDocument()` version:

      function makeDocument(resource, url) {
        // create a new HTML document
        var doc = document.implementation.createHTMLDocument(IMPORT_LINK_TYPE);
        // cache the document's source url
        doc._URL = url;
      
        // install html
        doc.body.innerHTML = resource;

        // Ensure an absolute URL
        if (!/^(https?:)?\/\//.test(url)) {
            var a = document.createElement('a');
            a.href = url;
            url = a.href;
        }
      
        // establish a relative path via <base>  
        var base = doc.querySelector('base');
        if (!base) {
            base = doc.createElement('base');
            base.setAttribute("href", url);
            doc.head.appendChild(base);
        }
        // add baseURI support to browsers (IE) that lack it.
        //if (!doc.baseURI && !hasBaseURIAccessor(doc)) {
          Object.defineProperty(doc, "baseURI", {
            value: url
          });
        //}

        // ensure UTF-8 charset
        var meta = doc.createElement("meta");
        meta.setAttribute("charset", "utf-8");
        doc.head.appendChild(meta);

        // TODO(sorvell): ideally this code is not aware of Template polyfill,
        // but for now the polyfill needs help to bootstrap these templates
        if (window.HTMLTemplateElement && HTMLTemplateElement.bootstrap) {
          HTMLTemplateElement.bootstrap(doc);
        }
        return doc;
      }

This sets the `<base>` tag only if one does not already exist.  However, I found that any components that were then imported by these components were called with a relative instead of absolute URL which appeared to break things when it was used as the `<base>` url here if they didn’t specify their own.  Therefore I added the check to always convert to an absolute URL.  Also, I commented out your check for an existing `baseURI`.  Components with a `<base>` set came in with `baseURI` set to `about:blank` in FireFox (not IE11), not sure how you want to handle that..

This patch appears to work for me with IE11.  However, it still does not work in FireFox with the following error in `register()`:

        if (getRegisteredDefinition(name)) {
          throw new Error("DuplicateDefinitionError: a type with name '" + String(name) + "' is already registered");
        }

I have a component `A` with a `<base>` tag which imports several components.  Some of these have `<base>` and some do not, let’s call both cases `B` and `C`.  Then both `B` and `C` import component `D`.  It appears that somewhere the webcomponents polyfill isn’t considering `D` the same in FireFox and trying to register it twice.  Thoughts?  I hope this is helpful for you and getting close.

**Comment by jmesserly** [comment]
Awesome investigation notes!!!

> [...] It appears that somewhere the webcomponents polyfill isn’t considering D the same in FireFox and trying to register it twice. Thoughts? I hope this is helpful for you and getting close.

Hmmm. That sounds like it's importing D twice, maybe because it's using relative instead of absolute URI for de-duping? The deduping is supposed to happen here:
https://github.com/webcomponents/webcomponentsjs/blob/51c337b2327ef63ee50ee1b69f50dc19966f1483/src/HTMLImports/Loader.js#L53

I wonder if we're getting two different URLs for D in firefox (but same URL in IE 11)

**Comment by drarmstr** [comment]
Tada!  Patching `require()` as follows fixes FireFox:

    require: function(elt) {
      var url = elt.src || elt.href;

      // Ensure an absolute URL
      if (!/^(https?:)?\/\//.test(url)) {
        var a = document.createElement('a');
        a.href = url;
        url = a.href;
      }
      
      // ensure we have a standard url that can be used
      // reliably for deduping.
      // TODO(sjmiles): ad-hoc
      elt.__nodeUrl = url;
      
      if (!this.dedupe(url, elt)) {
        // fetch this resource
        this.fetch(url, elt);
      }
    },

In fact, adding the relative->absolute URL conversion here means it doesn't have to be done in `makeDocument()`.  So, hopefully you'll consider this patch?

**Comment by jmesserly** [comment]
Definitely. Do you want to make it into a pull request?

one good thing to check, running the existing tests. Basically start a local server of some sort, then open this file: https://github.com/webcomponents/webcomponentsjs/blob/master/tests/runner.html
in whatever browser(s) you have access too.

CC @addyosmani as well

**Comment by drarmstr** [comment]
Ok, I can try to do this.  There is some legal bureaucracy, though...  Can you refer me to your corporate contributors agreement for our legal department?

**Comment by jmesserly** [comment]
Yup, it's here: https://cla.developers.google.com/about/google-corporate

**Comment by drarmstr** [comment]
I see some errors, but perhaps they are known failures or due to a bad environment as I see them before I make any changes:

* html/style-links.html
* html/style-import-base-tag.html
* html/ce-import-upgrade.html
* html/ce-import-upgrade-async.html
